lib/reversetunnel/track/tracker.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package track
+
+import (
+	"context"
+	//"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/workpool"
+)
+
+type Lease = workpool.Lease
+
+// Key uniquely identifies a reversetunnel endpoint.
+type Key struct {
+	Cluster string
+	Type    string
+	Addr    utils.NetAddr
+}
+
+type Config struct {
+	ProxyExpiry time.Duration
+	TickRate    time.Duration
+}
+
+func (c *Config) SetDefaults() {
+	if c.ProxyExpiry < 1 {
+		c.ProxyExpiry = time.Minute * 3
+	}
+	if c.TickRate < 1 {
+		c.TickRate = time.Second * 30
+	}
+}
+
+// Tracker is a helper for tracking proxies located behind reverse tunnels
+// and triggering agent spawning as needed.
+type Tracker interface {
+	// Start starts tracking for specified key.
+	Start(key Key)
+	// Stop stops tracking for specified key.
+	Stop(key Key)
+	// Acquire grants access to the Acquire channel of the
+	// embedded work group.
+	Acquire() <-chan Lease
+	// MarkSeen starts/refreshes tracking for supplied proxies.
+	MarkSeen(lease Lease, proxies ...string)
+	// WithProxy runs the supplied closure if and only if
+	// no other work is currently being done with the proxy
+	// identified by principals.
+	WithProxy(work func(), lease Lease, principals ...string) (didWork bool)
```

Comment:
```go
// WithProxy tries to acquire access to a proxy. If the proxy has already been claimed
// by another worker, will return right away. Otherwise will run the closure.
```